lib/hope-tabs.js

```diff
@@ -0,0 +1,722 @@
+    this.selecting = false;
+    this.opening = false;
+    this.closing = false;
+
+    // Events
+    scheduler.attachDirect(this.els.inner, 'scroll', this.onScroll);
+    on(this.els.container, 'close-clicked', e => this.onClose(e));
+    on(this.els.container, 'click', e => this.onClick(e));
+    on(this.els.add, 'click', e => this.onAddClick(e));
+
+    this.setupSnapping();
+
+    // Layout when decendants are ready
+    setTimeout(() => this.layout());
+  },
+
+  onClick(e) {
+    var tab = e.target.closest('li');
+    if (tab) this.select(tab);
+  },
+
+  onClose(e) {
+    var tab = e.target.closest('li');
+    this.close(tab);
+  },
+
+  onScroll() {
+    var height = window.innerHeight;
+    var previous = this.scrollPosition;
+    var current = this.getScrollTop();
+    var entering = current > previous && current >= height / 2;
+    var leaving = current < previous && current <= height / 2;
+
+    if (entering) this.onEnterTabs();
+    if (leaving) this.onLeaveTabs();
+
+    this.scrollPosition = current;
+  },
+
+  onEnterTabs() {
+    if (this.inTabsView) return;
+    this.inTabsView = true;
+    debug('enter tabs');
+
+    var current = this.els.tabs[0];
+    if (!current) return;
+
+    scheduler.feedback(() => {
+      this.els.grippy.style.opacity = 0;
+      current.firstElementChild.active = false;
+      this.els.add.classList.add('show');
+    }, current, 'transitionend');
+  },
+
+  onLeaveTabs() {
+    if (!this.inTabsView) return;
+    this.inTabsView = false;
+    debug('leave tabs');
+
+    var current = this.els.tabs[0];
+    if (!current) return;
+
+    scheduler.feedback(() => {
+      this.els.grippy.style.opacity = '';
+      this.els.add.classList.remove('show');
+      current.firstElementChild.active = true;
+    }, current, 'transitionend');
+  },
+
+  onAddClick() {
+    this.open({
+      title: 'Home',
+      url: 'homescreen.gaiamobile.org'
+    });
+  },
+
+  getSnapHeight() {
+    return window.innerHeight -
+      PREVIEW_HEIGHT -
+      GUTTER_HEIGHT -
+      BAR_HEIGHT;
+  },
+
+  getTabHeight() {
+    return Math.max(
+      TAB_HEIGHT,
+      this.getSnapHeight() / this.els.tabs.length
+    );
+  },
+
+  getScrollTop() {
+    return this.els.inner.scrollTop;
+  },
+
+  scrollTo(top, options={}) {
+    return new Promise(resolve => {
+      this.els.inner.scrollTo({
+        top: top,
+        behavior: options.instant ? 'auto' : 'smooth'
+      });
+
+      setTimeout(resolve, options.wait || 300);
+    });
+  },
+
+  setupSnapping() {
+    var snapHeight = this.getSnapHeight();
+    this.els.inner.style.scrollSnapPointsY = `repeat(${snapHeight}px)`;
+  },
+
+  layout() {
+    return new Promise(resolve => {
+      debug('layout');
+
+      var current = this.els.tabs[0];
+      var content = current.querySelector('hope-tab');
+      var snapHeight = this.getSnapHeight();
+      var container = this.els.container;
+
+      this.cleanUp(current);
+
+      content.active = !window.inTabsView;
```

Comment:
probably want `!this.inTabsView` here, otherwise the chrome shows up on the top tab when you close a tab below